subprojects/core/src/main/java/org/gradle/internal/operations/trace/BuildOperationTrace.java

```diff
@@ -112,9 +115,32 @@ public BuildOperationTrace(StartParameter startParameter, BuildOperationListener
         }
 
         listener = new SerializingBuildOperationListener(logOutputStream);
-        listenerManager.addListener(listener);
-    }
+        buildOperationListenerManager.addListener(listener);
+
+        // This is a workaround for https://github.com/gradle/gradle/issues/3873
+        // Several early typed operations have `buildPath` property,
+        // the value of which can only be determined after the settings file for the build has loaded.
+        //
+        // The workaround is to buffer all operation notifications in memory until the root build's settings have loaded.
+        // This works because all possible settings files have been evaluated by the time the root one has been.
+        // This is not guaranteed to hold into the future.
+        // A proper solution would be to change the operation details/results to be
+        // truly immutable and convey values known at the time.
+        listenerManager.addListener(new BuildAdapter() {
```

Comment:
❌ It seems this listener is never removed, which could lead to a memory leak.

Reply:
I'll have a look on this

Reply:
added a fix for this in https://github.com/gradle/gradle/pull/4278